tests: run more tests with all targets by default

JackieWeiss opened the pull request:
Summary
=======
Enabled all test targets for some tests, fixed tests where necessary.
Better test coverage for JS and VM.

Details
=======
- Reordered the target category strings to be in alphabetical order to
make later adjustments easier.
- The  `targets`  key is removed for all tests except one to disable a
target where the target doesn't make
   sense.
-  `write(stdout, ...)`  is replaced with  `echo ...`  to support the JS
and VM targets.
- Some  `static: `  blocks have been removed and replaced with the VM
target where not specific to compile-time execution.
- Tests that fail but should work are marked as `knownIssue`.

part of https://github.com/nim-works/nimskull/issues/1075

Commit:
tests: run more tests with all targets by default

Summary
=======
Enabled all test targets for some tests. Better test coverage for JS and
VM.

Details
=======
Reordered the target category strings to be in alphabetical order to make
later adjustments easier.
The 'targets' key is removed for all tests.
'write(stdout, ...)' is replaced with 'echo ...' to support the JS and
VM targets.
Some 'static: ' blocks have been removed and replaced with the VM target
where not specific to compile-time execution.
Tests that fail but should work are marked as 'knownIssue'.

zerbina commented on tests/assert/tassert_c.nim:
```diff
@@ -1,11 +1,16 @@
+  knownIssue.js vm:'''
+  JavaScript produces a different stacktrace output
+  message than the c target.
+  VM does not support 'getStackTrace' on an Exception.
```
```suggestion
    JavaScript produces a different stacktrace output message than the C
    target.
    VM does not support 'getStackTrace' on an Exception.
```

zerbina commented on tests/align/talign.nim:
```diff
@@ -1,6 +1,5 @@
 discard """
 ccodeCheck: "\\i @'NIM_ALIGN(128) NI mylocal1' .*"
-targets: "c"
```
The JS target has no concept of alignment, and I don't think it ever will.

You can use negative form, `targets: "!js"`, to disable only the JS target while still inheriting the remaining default targets.

zerbina commented on tests/assert/tassert_c.nim:
```diff
@@ -1,11 +1,16 @@
 discard """
   cmd: "nim $target $options --excessiveStackTrace:off $file"
```
```suggestion
  matrix: "--excessiveStackTrace:off"
```
If you come across a `cmd` specification that doesn't change the command that's run (e.g., `check`), you can replace it with a single-item `matrix`. This makes it clearer for both `testament` and the reader what's happening.

zerbina commented on tests/assign/tobject_assign.nim:
```diff
@@ -3,6 +3,9 @@ discard """
     has a type header
   '''
+  knownIssue.vm: '''
+  Cannot 'importc' proc 'localtime' in 'times.nim' at compile time
```
```suggestion
    'localtime' is not yet supported with the VM backend
```
The error message as reported by the compiler is misleading, mentioning compile-time while the problem is not at all related to compile-time evaluation.

**Edit:** I've replaced the backticks with single quotes.

zerbina replied:
That's a habit of mine from writing documentation comments, where names are wrapped in backticks.

For test specifications, I think it doesn't matter, although it should be locally consistent within a single `knownIssue` or `description` item.

zerbina commented on tests/assign/toverload_asgn2.nim:
```diff
@@ -1,7 +1,7 @@
   output: '''i value 88
 2aa'''
-    knownIssue: "possible issue where result is not treated as var Concrete in passAround"
+  knownIssue.c js vm: "possible issue where result is not treated as var Concrete in passAround"
```
```suggestion
  knownIssue: "possible issue where result is not treated as var Concrete in passAround"
```
Since the bug is unrelated to the selected backend (it's a bug in semantic analysis), keeping the `knownIssue` unspecific is a bit clearer.

zerbina reviewed (changes requested):
A very well-made PR. Thank you very much!

I left a few suggestions regarding the wording and styling of `knownIssue`s, as well as some additional requests.

Also, a small request regarding the PR description: could you use a listing for the items in the PR's 'Details' section? That'd make it a bit easier to parse.

Commit:
Fix style, targets and correct knownIssue message

Summary
=======
Fixed some minor issues with test specifications.
Changed a knownIssue message.

Detail
=======
. Changed indentation for a knownIssue message.
. Changed targets for a test where a target doesn't make sense.
. Changed knownIssue target where the issue is unrelated to the
  backends.
. Changed a knownIssue message to point to the actual error instead of
  the error the compiler emitted.
. Changed a 'cmd' to a 'matrix' to make intent clearer.

zerbina reviewed (approved):
Very good work, thanks again. Also, good call on reordering the strings in `defaultTargets`.

zerbina commented:
/merge
